### DISCUSSION TITLE
Can you pack and reuse pointers? Trying to make LibRealsense garbage free

### DISCUSSION THREAD

**Discussion by btshrewsbury** [bytedeco/javacpp#532] [General]
Hi Saudet!

I am trying to get depth data from a few real sense cameras into a real time java thread and I would like to avoid calling the GC if possible. I was able to profile and it looks like most of the allocations on the Java side are related to new pointer allocations (Pointer, rs2_frame, and rs2_stream_profile). Do you know if it's possible to reuse pointers instead of creating new ones each javacpp call? I think we can manually call the deallocation of native memory within the realtime thread once we are done with the byte buffers. I don't fully understand how the pointers get allocated and managed under the hood, so I am not really sure where to go from here. 

Javacpp is very helpful, Thank you!

**Comment by saudet** [comment]
If you're talking about deterministic deallocation, we can do that with Pointer.close(), but PointerScope is more convenient:
http://bytedeco.org/news/2018/07/17/bytedeco-as-distribution/
If you're talking about reusing native memory for RealSense, that's going to depend on its API and what it's capable of.
Or are you talking about something else?

**Comment by btshrewsbury** [comment]
I may not be describing it well so code might just be easier? I just don't want to call new. It looks like polling allows us to pack rs2_frame without allocating. rs2_get_frame_data requires a new BytePointer though. I just want to reuse it every tick. I was looking at manually calling deallocate, null, then init on the packablePointer, but I haven't tried it yet.


```
rs2_frame depth_frame = new rs2_frame();
BytePointer packablePointer= new BytePointer();

while (true) 
{
     realsense2.rs2_pipeline_poll_for_frames(pipe, depth_frame, e);

     //this is the way it's typically 
     BytePointer newInstantiatedPointer = new BytePointer(realsense2.rs2_get_frame_data(depth_frame, e)); done

      //this is what I want to do, so I only have to instantiate the byte pointer once. 
     packablePointer.set(realsense2.rs2_get_frame_data(depth_frame, e));

     //do stuff with the packed pointer data

     //allow the C side to release the data
     packablePointer.deallocate();
}
```

**Reply by saudet** [reply]
rs2_get_frame_data() also creates a new Pointer though. The data for the same frame shouldn't change, so you should be able to call rs2_get_frame_data() only once on a given rs2_frame, and then you should be able to reuse the data pointer you got for the frame over and over again. That's how we do this kind of thing.

**Reply by btshrewsbury** [reply]
Hmmm. It looks like realsense2.rs2_get_frame_data(depth_frame, e) isn't easy to get around either. Yourkit isn't great at JNI calls. I was hoping the allocation was from the outer instantiation, and the rs2_get_frame_data was passing a reference from the C side. 

Do you think it would be possible to create realsense2.rs2_get_frame_data(depth_frame, bytePointer, e) that's allocation free? I had expected I would need to write some C stuff and wrap it with JNI, but  javacpp is almost exactly what we need

**Reply by saudet** [reply]
I don't think rs2_get_frame_data() allocates memory, it probably just returns the same pointer over and over again, so just call it once, if that's the case.

**Reply by btshrewsbury** [reply]
![image](https://user-images.githubusercontent.com/2991137/148491173-9c4f7e75-b4a1-45e9-aa69-8cd2fe1b942c.png)


I profiled the call by itself and it seems to create a javacpp.Pointer under the hood. I know it seems like overkill to try and remove it, but we are running a garbage free realtime java controller for our humanoid robot's and small allocations cause GC's over time. The GC can cause weird stuff to happen, usually just fall. Here's a video if you are interested: [https://www.youtube.com/watch?v=PqLZP8TANlg](https://www.youtube.com/watch?v=PqLZP8TANlg).

**Reply by saudet** [reply]
Yes, it will create a new Pointer object, but its address probably doesn't change! If you are simply getting the same address over and over again, do not call rs2_get_frame_data(), and reuse an old instance of that Pointer. If it's the same address, it will work fine.

**Reply by btshrewsbury** [reply]
I didn't understand at first, but that's a pretty ingenious hack. It looks like the pointer address does cycle: The first address is unique, then they repeat after that. I'll work on some way to probe the memory to make sure it's valid, otherwise I get an EXCEPTION_ACCESS_VIOLATION and it crashes hard. thank you very much for the help!

org.bytedeco.javacpp.MutableBytePointer[address=0x2ede7060,position=0,limit=0,capacity=0,deallocator=null]
org.bytedeco.javacpp.MutableBytePointer[address=0x2fc28060,position=0,limit=0,capacity=0,deallocator=null]
org.bytedeco.javacpp.MutableBytePointer[address=0x2f90a060,position=0,limit=0,capacity=0,deallocator=null]
org.bytedeco.javacpp.MutableBytePointer[address=0x2fc28060,position=0,limit=0,capacity=0,deallocator=null]
org.bytedeco.javacpp.MutableBytePointer[address=0x2f90a060,position=0,limit=0,capacity=0,deallocator=null]
org.bytedeco.javacpp.MutableBytePointer[address=0x2fc28060,position=0,limit=0,capacity=0,deallocator=null]
org.bytedeco.javacpp.MutableBytePointer[address=0x2f90a060,position=0,limit=0,capacity=0,deallocator=null]
org.bytedeco.javacpp.MutableBytePointer[address=0x2fc28060,position=0,limit=0,capacity=0,deallocator=null]
org.bytedeco.javacpp.MutableBytePointer[address=0x2f90a060,position=0,limit=0,capacity=0,deallocator=null]
org.bytedeco.javacpp.MutableBytePointer[address=0x2fc28060,position=0,limit=0,capacity=0,deallocator=null]
org.bytedeco.javacpp.MutableBytePointer[address=0x2f90a060,position=0,limit=0,capacity=0,deallocator=null]
org.bytedeco.javacpp.MutableBytePointer[address=0x2fc28060,position=0,limit=0,capacity=0,deallocator=null]
 

Also MutableBytePointer just extends ByteBuffer with the method set:

```
public void set(final Pointer p)
   {
      address = p.address;
      position = p.position;
      limit = p.limit;
      capacity = p.capacity;
      
      if(p.deallocator() != null)
      {
         deallocator(new ProxyDeallocator(this, p));
      }
   }
```

**Reply by saudet** [reply]
There is most likely a one-to-one correspondence between the rs2_frame itself and what rs2_frame_data() returns...

**Reply by saudet** [reply]
I've added an `rs2_get_frame_data_address()` overload that doesn't create new objects and returns the address as a `long` in commit https://github.com/bytedeco/javacpp-presets/commit/7f0122b9c3c06eea74965870a86732e2600094ad. You can use that value to set the address of, for example, your MutableBytePointer. Please give it a try with the snapshots: http://bytedeco.org/builds/

**Reply by btshrewsbury** [reply]
It works great, Thank you! 

Yourkit is complaining about depth_pointer.get(depth_bytes, 0, depth_data_size) now. It looks like just about every pointer method returns a new pointer so you can perform nice method chaining. I tried manually calling get() while incrementing the offset to go through the bytes manually, but that calls a method under the hood that also creates a pointer. Is there a way to get the contents of a byte array without creating a new pointer? Sorry to be a pain in the ass. 

I tried to use asBuffer(), but the result ByteBuffer has a limit of 1 and a capacity of 1.

**Reply by saudet** [reply]
You mean this method? That doesn't create any new objects, no.
http://bytedeco.org/javacpp/apidocs/org/bytedeco/javacpp/BytePointer.html#get-byte:A-int-int-

What you want to use though is probably an indexer:
http://bytedeco.org/news/2014/12/23/third-release/

**Reply by btshrewsbury** [reply]
Using JVM statistics the memory usage stays constant so everything is working. No idea what Yourkit thinks is going on.  Thank you again for the help! I will look at the indexer to help speed up moving the data around.

**Reply by saudet** [reply]
Never fully trust what a profiler tells you :)
